File: backend/app/schemas/quality.py
"""运营质量指标的公开响应契约。"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


VersionDimension = Literal[
    "model",
    "prompt_digest",
    "rules_digest",
    "data_digest",
]


class NodeLatencyMetrics(BaseModel):
    samples: int = Field(ge=1)
    p50_ms: int = Field(ge=0)
    p95_ms: int = Field(ge=0)


class GenerationQualityMetrics(BaseModel):
    total: int = Field(ge=0)
    completed: int = Field(ge=0)
    failed: int = Field(ge=0)
    cancelled: int = Field(ge=0)
    active: int = Field(ge=0)
    success_rate: float | None = Field(default=None, ge=0, le=1)
    fallback_rate: float | None = Field(default=None, ge=0, le=1)
    duration_p50_ms: int | None = Field(default=None, ge=0)
    duration_p95_ms: int | None = Field(default=None, ge=0)
    total_tokens: int = Field(ge=0)
    total_cost_cny: float = Field(ge=0)
    node_latency: dict[str, NodeLatencyMetrics]


class GenerationVersionCohortItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    model: str | None = None
    prompt_digest: str | None = None
    rules_digest: str | None = None
    data_digest: str | None = None
    version_complete: bool
    missing_dimensions: list[VersionDimension]
    total: int = Field(ge=0)
    completed: int = Field(ge=0)
    failed: int = Field(ge=0)
    cancelled: int = Field(ge=0)
    active: int = Field(ge=0)
    success_rate: float | None = Field(default=None, ge=0, le=1)
    fallback_rate: float | None = Field(default=None, ge=0, le=1)
    duration_p50_ms: int | None = Field(default=None, ge=0)
    duration_p95_ms: int | None = Field(default=None, ge=0)
    total_tokens: int = Field(ge=0)
    known_cost_cny: float = Field(ge=0)
    unknown_cost_run_count: int = Field(ge=0)


class GenerationVersionCohorts(BaseModel):
    total_cohorts: int = Field(ge=0)
    returned_cohorts: int = Field(ge=0)
    truncated: bool
    items: list[GenerationVersionCohortItem]


class AgentQualityMetrics(BaseModel):
    turn_total: int = Field(ge=0)
    handoff_total: int = Field(ge=0)
    handoff_rate: float | None = Field(default=None, ge=0, le=1)
    statuses: dict[str, int]


class ModelCallQualityMetrics(BaseModel):
    total: int = Field(ge=0)
    succeeded: int = Field(ge=0)
    failed: int = Field(ge=0)
    blocked: int = Field(ge=0)
    total_tokens: int = Field(ge=0)
    known_actual_cost_cny: float = Field(ge=0)
    unknown_cost_call_count: int = Field(ge=0)
    provider_failures: dict[str, int]


class LayoutQualityMetrics(BaseModel):
    total: int = Field(ge=0)
    hard_pass_total: int = Field(ge=0)
    hard_pass_rate: float | None = Field(default=None, ge=0, le=1)
    average_score: float | None = None
    issue_codes: dict[str, int]


class FeedbackActionCounts(BaseModel):
    adopt: int = Field(ge=0)
    remove: int = Field(ge=0)
    replace: int = Field(ge=0)
    move: int = Field(ge=0)
    final_select: int = Field(ge=0)


class FeedbackQualityMetrics(BaseModel):
    total: int = Field(ge=0)
    action_counts: FeedbackActionCounts
    modification_total: int = Field(ge=0)
    modification_rate: float | None = Field(default=None, ge=0, le=1)
    final_select_total: int = Field(ge=0)
    satisfaction_count: int = Field(ge=0)
    satisfaction_mean: float | None = Field(default=None, ge=1, le=5)
    glb_load_failure_total: int = Field(ge=0)


class RenderQueueQualityMetrics(BaseModel):
    total: int = Field(ge=0)
    queued: int = Field(ge=0)
    running: int = Field(ge=0)
    completed: int = Field(ge=0)
    failed: int = Field(ge=0)
    dead_letter: int = Field(ge=0)
    cancelled: int = Field(ge=0)
    success_rate: float | None = Field(default=None, ge=0, le=1)
    queue_wait_p50_ms: int | None = Field(default=None, ge=0)
    queue_wait_p95_ms: int | None = Field(default=None, ge=0)
    execution_p50_ms: int | None = Field(default=None, ge=0)
    execution_p95_ms: int | None = Field(default=None, ge=0)


class QualitySummaryResponse(BaseModel):
    generated_at: datetime
    window_days: int = Field(ge=1, le=365)
    generation: GenerationQualityMetrics
    version_cohorts: GenerationVersionCohorts
    agent: AgentQualityMetrics
    model_calls: ModelCallQualityMetrics
    layout: LayoutQualityMetrics
    feedback: FeedbackQualityMetrics
    effect_render: RenderQueueQualityMetrics
    blender_render: RenderQueueQualityMetrics
    failure_codes: dict[str, int]

File: frontend/src/types/quality.ts
export type QualityWindowDays = 7 | 30 | 90;

export interface QualityGenerationMetrics {
  total: number;
  completed: number;
  failed: number;
  cancelled: number;
  active: number;
  success_rate: number | null;
  fallback_rate: number | null;
  duration_p50_ms: number | null;
  duration_p95_ms: number | null;
  total_tokens: number;
  total_cost_cny: number;
  node_latency: Record<string, {
    samples: number;
    p50_ms: number;
    p95_ms: number;
  }>;
}

export interface QualityRenderQueueMetrics {
  total: number;
  queued: number;
  running: number;
  completed: number;
  failed: number;
  dead_letter: number;
  cancelled: number;
  success_rate: number | null;
  queue_wait_p50_ms: number | null;
  queue_wait_p95_ms: number | null;
  execution_p50_ms: number | null;
  execution_p95_ms: number | null;
}

export interface QualityAgentMetrics {
  turn_total: number;
  handoff_total: number;
  handoff_rate: number | null;
  statuses: Record<string, number>;
}

export interface QualityLayoutMetrics {
  total: number;
  hard_pass_total: number;
  hard_pass_rate: number | null;
  average_score: number | null;
  issue_codes: Record<string, number>;
}

export interface QualityFeedbackMetrics {
  total: number;
  action_counts: {
    adopt: number;
    remove: number;
    replace: number;
    move: number;
    final_select: number;
  };
  modification_total: number;
  modification_rate: number | null;
  final_select_total: number;
  satisfaction_count: number;
  satisfaction_mean: number | null;
  glb_load_failure_total: number;
}

export interface QualitySummary {
  generated_at: string;
  window_days: number;
  generation: QualityGenerationMetrics;
  agent: QualityAgentMetrics;
  layout: QualityLayoutMetrics;
  feedback: QualityFeedbackMetrics;
  effect_render: QualityRenderQueueMetrics;
  blender_render: QualityRenderQueueMetrics;
  failure_codes: Record<string, number>;
}

export type FailureSeverity = "low" | "medium" | "high" | "critical";
export type FailureStatus = "open" | "in_progress" | "resolved" | "verified";

export interface FailureCluster {
  id: number;
  fingerprint: string;
  taxonomy_version: string;
  data_version: string;
  failure_type: string;
  code: string;
  severity: FailureSeverity;
  status: FailureStatus;
  owner: string | null;
  occurrence_count: number;
  affected_count: number;
  first_seen_at: string;
  last_seen_at: string;
  detected_version: string;
  fixed_version: string | null;
  verified_version: string | null;
  created_at: string;
  updated_at: string;
}

export interface FailureClusterListResponse {
  items: FailureCluster[];
  summary: {
    total: number;
    by_status: Record<string, number>;
    by_severity: Record<string, number>;
  };
}

export interface FailureClusterUpdate {
  status?: FailureStatus;
  owner?: string | null;
  fixed_version?: string | null;
  verified_version?: string | null;
}

export interface FailureTriageReport {
  schema_version: "1.0";
  report_id: string;
  taxonomy_version: string;
  data_version: string;
  candidate_version: string;
  signature_algorithm: "hmac-sha256";
  signature_key_id: string;
  signature: string;
  generated_at: string;
  failures: Array<{
    failure_type: string;
    code: string;
    severity: FailureSeverity;
    occurrence_count: number;
    affected_count: number;
  }>;
}

export interface FailureTriageSyncResponse {
  imported: boolean;
  cluster_count: number;
  clusters: FailureCluster[];
}
